proxy/envoy/config.go

```diff
@@ -121,83 +112,126 @@ func Generate(instances []*model.ServiceInstance, services []*model.Service, mes
-		if !ok {
-			protocols = make(map[model.Protocol]bool)
-			ports[instance.Endpoint.Port.Port] = protocols
+func buildListeners(instances []*model.ServiceInstance, services []*model.Service) ([]Listener, []Cluster) {
+	clusters := buildClusters(services)
+	listeners := make([]Listener, 0)
+
+	// group by port values
+	type listener struct {
+		instances map[model.Protocol]bool
+		// services map service, port for a given port, protocol pair
+		services map[model.Protocol][]*model.Service
+	}
+
+	ports := make(map[int]*listener, 0)
+	ensure := func(port model.Port) {
+		if _, ok := ports[port.Port]; !ok {
+			ports[port.Port] = &listener{
+				instances: make(map[model.Protocol]bool),
+				services:  make(map[model.Protocol][]*model.Service),
+			}
 		}
-		protocols[instance.Endpoint.Port.Protocol] = true
 	}
 
-	listeners := make([]Listener, 0)
-	clusters := make([]Cluster, 0)
-	for port, protocols := range ports {
-		cluster := Cluster{
-			Name:             fmt.Sprintf("%s_%d", IngressClusterPrefix, port),
-			Type:             "static",
-			ConnectTimeoutMs: 1000,
-			LbType:           LbTypeRoundRobin,
-			Hosts:            []Host{{URL: fmt.Sprintf("tcp://%s:%d", "127.0.0.1", port)}},
+	for _, instance := range instances {
+		port := instance.Endpoint.Port
+		ensure(port)
+		ports[port.Port].instances[port.Protocol] = true
+	}
+
+	for _, svc := range services {
+		for _, port := range svc.Ports {
+			ensure(port)
+			ports[port.Port].services[port.Protocol] = append(ports[port.Port].services[port.Protocol], &model.Service{
+				Name:      svc.Name,
+				Namespace: svc.Namespace,
+				Ports:     []model.Port{port},
+			})
 		}
+	}
 
+	for port, lst := range ports {
 		listener := Listener{
 			Port:       port,
 			BindToPort: false,
 		}
 
+		ingressClusterName := fmt.Sprintf("%s%d", IngressClusterPrefix, port)
+		if len(lst.instances) > 0 {
+			clusters = append(clusters, Cluster{
+				Name:             ingressClusterName,
+				Type:             "static",
+				ConnectTimeoutMs: 1000,
+				LbType:           LbTypeRoundRobin,
+				Hosts:            []Host{{URL: fmt.Sprintf("tcp://%s:%d", "127.0.0.1", port)}},
+			})
+		}
+
 		// These two filters operate at different levels of the network stack.
 		// In practice, no port has two protocols used in the same set of instances, but we
 		// should be careful with not stepping on our feet.
-		if protocols[model.ProtocolTCP] {
+		// The order of the filter insertion is important.
+		if lst.instances[model.ProtocolTCP] {
 			listener.Filters = append(listener.Filters, NetworkFilter{
 				Type: "read",
 				Name: "tcp_proxy",
 				Config: NetworkFilterConfig{
-					Cluster:    cluster.Name,
+					Cluster:    ingressClusterName,
 					StatPrefix: "ingress_tcp",
 				},
 			})
 		}
 
-		if protocols[model.ProtocolGRPC] || protocols[model.ProtocolHTTP2] || protocols[model.ProtocolHTTP] {
+		// HTTP-level filter is shared between ingress and egress traffic.
```

Comment:
We should also add the options (not necessarily now) for the client to do TLS (terminated at Envoy) and/or use PROXY proto. These translate to different listener configs.